Lower std::string's alignment requirement from 16 to 8.

EricWF opened the pull request:
This allows smaller allocations to occur, closer to the actual std::string's required size. This is particularly effective in decreasing the allocation size upon initial construction (where __recommend is called to determine the size).

Although the memory savings per-string are never more than 8 bytes per string initially, this quickly adds up. And has lead to not insigficant memory savings at Google.

Unfortunately, this change is ABI breaking because it changes the value returned by max_size. So it has to be guarded.

EricWF commented:
This is a duplicate of #68749 which was closed by accident.

mordante commented on libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp:
```diff
@@ -0,0 +1,43 @@
+
+#include "test_macros.h"
+
+// alignment of the string heap buffer is hardcoded to 16
```
This comment seems wrong, the value 16 is no longer hard-coded.

EricWF replied:
Done.

mordante reviewed (commented):
In general no objection.
We should add this to the release notes too.

EricWF commented on libcxx/test/libcxx/strings/basic.string/string.capacity/max_size.pass.cpp:
```diff
@@ -18,7 +18,13 @@
-static const std::size_t alignment = 16;
+
+static const std::size_t alignment =
+#ifdef _LIBCPP_ABI_STRING_8_BYTE_ALIGNMENT
```
I would rather revisit this after landing the change.
There's more risk and more reason to reject the changes if we apply it globally.
So let's get this in, then consider if we can push it to everybody.

mordante replied:
I'm fine with that too, as mentioned I was mainly curious.

ldionne replied:
Can you prepare the patch to clean up the ABI flag and make it the default value before we merge this (kind of like a stacked review even though those aren't great in GH PRs)? I'd really like to make sure that we don't just check in this ABI flag without its removal if we all agree that we don't want the ABI flag in the end.

EricWF replied:
It was easier for me to write that patch on top  of HEAD, rather than this change. But here it is as requested.
#68925

EricWF commented:
OK, I have no idea how to rebase diffs apparently. Working on fixing this.

EricWF commented:
> In general no objection. We should add this to the release notes too.

Sorry, I wasn't ignoring this. But the release note should be there now.

ldionne commented on libcxx/include/__config:
```diff
@@ -167,6 +167,11 @@
 // The implementation moved to the header, but we still export the symbols from
 // the dylib for backwards compatibility.
 #    define _LIBCPP_ABI_DO_NOT_EXPORT_TO_CHARS_BASE_10
+// Same memory by providing the allocator more freedom to allocate the most
```
```suggestion
// Save memory by providing the allocator more freedom to allocate the most
```

EricWF replied:
lol sorry this snuck back in after I broke broke the last PR.

ldionne commented on libcxx/docs/ReleaseNotes/18.rst:
```diff
@@ -133,6 +133,13 @@ ABI Affecting Changes
+  inside std::string has decreased from 16 to 8 This save memory since string requests fewer additional
+  bytes than it did previously. However, this also changes the return value of std::string::max_length
+  and can cause code compiled against older libc++ versions but linked at runtime to a new version
+  to thrown a different exception when attempting allocations that are too large
```
```suggestion
  to throw a different exception when attempting allocations that are too large
```

EricWF replied:
Done.

ldionne commented on libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp:
```diff
@@ -0,0 +1,43 @@
+    assert(test_string.capacity() == expected_align8_size);
+  } else {
+    assert(test_string.capacity() == expected_align8_size + 8);
+  }
```
Can you explicitly `return 0` from the test? Otherwise this fails under `-ffreestanding` because `main` isn't a special function then.

EricWF replied:
Done.

ldionne reviewed (approved):
Thanks, this LGTM with a few nitpicks. The C++03 build is failing but I think this is easy to address (probably your usage of `constexpr` in the test).

mordante commented on libcxx/docs/ReleaseNotes/18.rst:
```diff
@@ -133,6 +133,13 @@ ABI Affecting Changes
   extremely rare. Any error resulting from this change should result in a link-time error.
 
+- Under the unstable ABI, the internal alignment requirements for heap allocations
+  inside std::string has decreased from 16 to 8 This save memory since string requests fewer additional
```
s/std::string/``std::string``/ so it will be rendered properly, the same for other places.

mordante commented on libcxx/docs/ReleaseNotes/18.rst:
```diff
@@ -133,6 +133,13 @@ ABI Affecting Changes
 
+- Under the unstable ABI, the internal alignment requirements for heap allocations
+  inside std::string has decreased from 16 to 8 This save memory since string requests fewer additional
+  bytes than it did previously. However, this also changes the return value of std::string::max_length
```
s/std::string::max_length/``std::string::max_size``

mordante commented:
> > In general no objection. We should add this to the release notes too.
> 
> Sorry, I wasn't ignoring this. But the release note should be there now.

No problem.

mordante commented:
sorry wrong button, still getting used to GitHub reviews

mordante reviewed (approved):
LGTM modulo some nits.

EricWF committed:
Lower std::string's alignment requirement from 16 to 8

This allows smaller allocations to occur, closer to the actual std::string's required size. This is particularly effective in decreasing the allocation size upon initial construction (where __recommend is called to determine the size).

Although the memory savings per-string are never more than 8 bytes per string initially, this quickly adds up. And has lead to not insigficant memory savings at Google.

Unfortunately, this change is ABI breaking because it changes the value returned by max_size. So it has to be guarded.
